_posts/2024-02-23-enhanced-multi-vector-support-in-opensearch-knn.md

```diff
@@ -0,0 +1,182 @@
+
+Consider an illustrative example to delve into the issue further. Suppose there are 7 indexed documents, each containing either one or two vectors. Let's assume that, when searching the nested vector field, the order of vectors from nearest to farthest is as follows: Vector A, B, C, D, E, F, G, H, I, and J. The search occurs at the segment level, and with a k value of 2, only document 1 is returned from segment 1. Consequently, the final aggregated result comprises document 1 and document 3, whereas the expected outcome should include document 1 and document 2.
+
+<img src="/assets/media/blog-images/2024-02-23-multi-vector-support-in-knn/multi-vector-before-2.png" alt="multi-vector-behavior-before-cluster-level" class="img-centered"/>
+
+## Improvements in OpenSearch 2.12
+In the latest version, OpenSearch leverages document and vector mapping data to deduplicate search results. This process occurs when vectors belonging to the same document are already collected in the search queue. During deduplication, OpenSearch selects the vector closest to the query vector and discards the farther one. This value is then converted back to an OpenSearch search score, ultimately becoming the document's score.
+
+Let’s take a look at an example with the diagrams below. Consider two documents, Document 1 and Document 2. Document 1 contains vectors A and B, while Document 2 contains vectors C and D (Fig. 1). During the search, vector A is found and added to the search queue (Fig. 2). Subsequently, vector B is encountered. Since both vectors A and B belong to Document 1, their distances are compared. The distance between the query vector and vector B is 0.9, which is less than the distance between the query vector and vector A, which is 1.0. As a result, vector A is removed from the search queue, and vector B is added (Fig. 3).
+
+Moving forward, vector D is discovered. Since vector D belongs to Document 2, whose vector is not in the search queue, it is added (Fig. 4). Then, vector C is found. As vector C belongs to Document 2, whose vector is already in the search queue, the distances between the query vector and vectors C and D are compared. The distance between the query vector and vector D (1.1) is less than that between the query vector and vector C (1.2), so vector D remains in the search queue (Fig. 5).
```

Comment:
```suggestion
After that, vector D is discovered. Since vector D belongs to Document 2, whose vector is not in the search queue, vector D is added to the queue (Fig. 4). Then, vector C is found. Because vector C belongs to Document 2, whose vector is already in the search queue, the distances between the query vector and vectors C and D are compared. The distance between the query vector and vector D (1.1) is less than the distance between the query vector and vector C (1.2), so vector D remains in the search queue (Fig. 5).
```